### PR TITLE
Use ghcup to install GHC on linux and remove docs and profiling libs

### PR DESCRIPTION
**Pull request by hasufell**
See https://github.com/actions/virtual-environments/pull/4016 for discussion and reasoning

### CODE REVIEW

File: images/linux/scripts/installers/haskell.sh [hunk 1]
```diff
@@ -25,18 +25,24 @@ curl --proto '=https' --tlsv1.2 -sSf https://get-ghcup.haskell.org | sh > /dev/n
...
+    ghcup set ghc $fullVersion
+
+    # remove docs and profiling libs
+    ghc_bin_dir="$(ghcup whereis --directory ghc $fullVersion)"
+    [ -e "${ghc_bin_dir}" ] || exit 1
+    ghc_dir="$(realpath "${ghc_bin_dir}"/..)"
+    [ -e "${ghc_dir}" ] || exit 1
+    find "${ghc_dir}" \( -name "*_p.a" -o -name "*.p_hi" \) -type f -delete
+    rm -r "${ghc_dir}"/share/*
+    unset ghc_bin_dir ghc_dir
```
Comment (hasufell) [inline images/linux/scripts/installers/haskell.sh:45]: This part can and should be added to macos too.

Comment (miketimofeev) [reply images/linux/scripts/installers/haskell.sh:45]: there is no `realpath` util on macos by default, unfortunately, so we have to rework it a bit

Comment (hasufell) [reply images/linux/scripts/installers/haskell.sh:45]: You can get inspiration here:

https://github.com/haskell/ghcup/blob/8f4705a50eefafd92ceef32094eec7140885b3d5/ghcup#L1176-L1214

which is pure posix shell.

**Comment by miketimofeev** [comment]
@hasufell it looks like we don't need these
https://github.com/actions/virtual-environments/blob/3a8831b648dc326403ca2b02da9f98db7154bc7b/images/linux/scripts/installers/haskell.sh#L17-L22
and these lines anymore
https://github.com/actions/virtual-environments/blob/3a8831b648dc326403ca2b02da9f98db7154bc7b/images/linux/scripts/installers/haskell.sh#L51-L54

**Comment by hasufell** [comment]
@miketimofeev fixed

**Comment by hasufell** [comment]
https://github.visualstudio.com/virtual-environments/_build/results?buildId=116192&view=logs&j=011e1ec8-6569-5e69-4f06-baf193d1351e&t=5431112d-2b61-5a5f-7042-ef698f761043&l=6504

We don't install into /opt/ghc anymore.

**Comment by miketimofeev** [comment]
> https://github.visualstudio.com/virtual-environments/_build/results?buildId=116192&view=logs&j=011e1ec8-6569-5e69-4f06-baf193d1351e&t=5431112d-2b61-5a5f-7042-ef698f761043&l=6504
> 
> We don't install into /opt/ghc anymore.

yeah, could you please fix the test?
https://github.com/actions/virtual-environments/blob/main/images/linux/scripts/tests/Haskell.Tests.ps1

**Comment by hasufell** [comment]
I removed the directory tests.

**Comment by miketimofeev** [comment]
@hasufell sorry, forgot about one more place to change
https://github.com/actions/virtual-environments/blob/baabb70137a0d37b80d833010c4c3bb9ba84d207/images/linux/scripts/SoftwareReport/SoftwareReport.Common.psm1#L278-L279
we don't need `$aptSourceRepo` and  `(apt source repository: $aptSourceRepo)` anymore

**Comment by hasufell** [comment]
Fixed

**Review by miketimofeev — DISMISSED** [review]

File: images/linux/scripts/installers/haskell.sh [hunk 2]
```diff
@@ -14,29 +14,29 @@ export BOOTSTRAP_HASKELL_GHC_VERSION=0
...
-for version in $ghcMajorMinorVersions; do
-    # Get latest patch version for given Major.Minor one (ex. 8.6.5 for 8.6) and install it
-    exactVersion=$(echo "$allGhcVersions" | grep $version | sort --unique --version-sort | tail -1)
-    apt-get install -y ghc-$exactVersion
-    ghcInstalledVersions+=("$exactVersion")
-    defaultGHCVersion=$exactVersion
+availableVersions=$(ghcup list -t ghc -r | grep -v "prerelease" | awk '{print $2}')
+
+# Install 3 latest major versions(For instance 8.6.5, 8.8.4, 8.10.2)
+minorMajorVersions=$(echo "$availableVersions" | cut -d"." -f 1,2 | uniq | tail -n3)
```
Comment (al-cheb) [inline images/linux/scripts/installers/haskell.sh:25]: @hasufell , Currently, we install the last 2 major version. Could you please remove one?

```
allGhcVersions=$(apt-cache search "^ghc-" | grep -Po '(\d*\.){2}\d*' | sort --unique --version-sort)
echo "$allGhcVersions" | cut -d "." -f 1,2 | sort --unique --version-sort | tail -2
8.10
9.0
```

**Review by al-cheb — CHANGES_REQUESTED** [review]

File: images/linux/scripts/tests/Haskell.Tests.ps1 [hunk 3]
```diff
@@ -1,22 +1,5 @@
 Describe "Haskell" {
```
Comment (al-cheb) [inline images/linux/scripts/tests/Haskell.Tests.ps1:2]: @hasufell , Could you please revert and fix these tests? We need to be sure that we have 2 version have been installed and we can get --version output for them.

- GHC directory contains two version of GHC
- GHC version

**Comment by hasufell** [comment]
Should be done

**Review by miketimofeev — APPROVED** [review]

**Review by al-cheb — APPROVED** [review]

**Comment by miketimofeev** [comment]
@hasufell thanks for your contribution! The changes will be available next week.

**Pull request merged**